File: VB/e4643VB/Controllers/HomeController.vb
Public Class HomeController
    Inherits System.Web.Mvc.Controller
    Private list As New PersonList()
    Function Index() As ActionResult
        Return View(list.GetPersons())
    End Function

	Public Function GridViewPartial() As ActionResult
		Return PartialView("_GridViewPartial", list.GetPersons())
	End Function

	Public Function CustomGridViewEditingPartial(ByVal key As Integer) As ActionResult
		ViewData("key") = key
		Return PartialView("_GridViewPartial", list.GetPersons())
	End Function

	<HttpPost, ValidateInput(False)>
	Public Function EditingAddNew(ByVal person As Person) As ActionResult
		If ModelState.IsValid Then
			list.AddPerson(person)
		End If
		Return PartialView("_GridViewPartial", list.GetPersons())
	End Function

	<HttpPost, ValidateInput(False)>
	Public Function EditingUpdate(ByVal personInfo As Person) As ActionResult
		If ModelState.IsValid Then
			list.UpdatePerson(personInfo)
		End If
		Return PartialView("_GridViewPartial", list.GetPersons())
	End Function

	Public Function EditingDelete(ByVal personId As Integer) As ActionResult
		list.DeletePerson(personId)
		Return PartialView("_GridViewPartial", list.GetPersons())
	End Function
End Class

File: CS/e4643/Controllers/HomeController.cs
using e4643.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace e4643.Controllers
{
    public class HomeController : Controller
    {
        PersonList list = new PersonList();
        public ActionResult Index()
        {
            return View(list.GetPersons());
        }

        public ActionResult GridViewPartial()
        {
            return PartialView("_GridViewPartial", list.GetPersons());
        }

        public ActionResult CustomGridViewEditingPartial(int key)
        {
            ViewData["key"] = key;
            return PartialView("_GridViewPartial", list.GetPersons());
        }

        [HttpPost, ValidateInput(false)]
        public ActionResult EditingAddNew(Person person)
        {
            if (ModelState.IsValid)
                list.AddPerson(person);
            return PartialView("_GridViewPartial", list.GetPersons());
        }

        [HttpPost, ValidateInput(false)]
        public ActionResult EditingUpdate(Person personInfo)
        {
            if (ModelState.IsValid)
                list.UpdatePerson(personInfo);
            return PartialView("_GridViewPartial", list.GetPersons());
        }

        public ActionResult EditingDelete(int personId)
        {
            list.DeletePerson(personId);
            return PartialView("_GridViewPartial", list.GetPersons());
        }
    }
}